File: tests/fakes/fake_import_graph.py

from dataclasses import dataclass, field

from chisel.checker.models.import_edge import ImportEdge
from chisel.checker.models.layer import Layer


@dataclass(slots=True)
class FakeImportGraph:
    _edges: list[ImportEdge] = field(default_factory=list)
    _layer_map: dict[str, Layer] = field(default_factory=dict)

    def build(self, *args: object, **kwargs: object) -> None:
        return

    @property
    def all_imports(self) -> list[ImportEdge]:
        return list(self._edges)

    def module_layer(self, module_name: str, package_name: str) -> Layer | None:
        return self._layer_map.get(module_name)

    def set_edges(self, edges: list[ImportEdge]) -> None:
        self._edges = list(edges)

    def set_layers(self, module_name: str, layer: Layer) -> None:
        self._layer_map[module_name] = layer

File: chisel_js/tests/fakes/fake_import_graph.ts
import type { IImportGraph } from "chisel/checker/repositories/protocols";
import type { ImportEdge } from "chisel/checker/models/import_edge";
import { createImportEdge } from "chisel/checker/models/import_edge";

/**
 * Hand-fed import graph for service tests, mirroring
 * `chisel_py/tests/fakes/fake_import_graph.py`. Lets a boundary rule be tested
 * on an exact edge set without a project tree on disk.
 */
export class FakeImportGraph implements IImportGraph {
  private _edges: ImportEdge[] = [];
  readonly warnings: readonly string[] = [];

  build(): void {
    // Edges are supplied directly by the test.
  }

  get allImports(): readonly ImportEdge[] {
    return this._edges;
  }

  get unresolved(): readonly ImportEdge[] {
    return this._edges.filter(e => !e.resolved);
  }

  setEdges(edges: ImportEdge[]): void {
    this._edges = edges;
  }
}

/** Build an edge with the fields a test cares about; the rest get sane defaults. */
export function edge(params: {
  importer: string;
  imported: string;
  specifier?: string;
  isInternal?: boolean;
  resolved?: boolean;
  isTypeOnly?: boolean;
  lineNumber?: number;
  lineContents?: string;
}): ImportEdge {
  return createImportEdge({
    importer: params.importer,
    imported: params.imported,
    specifier: params.specifier ?? params.imported,
    isInternal: params.isInternal ?? params.imported.startsWith("src/"),
    resolved: params.resolved ?? true,
    isTypeOnly: params.isTypeOnly ?? false,
    lineNumber: params.lineNumber ?? 1,
    lineContents: params.lineContents ?? `import { x } from "${params.specifier ?? params.imported}";`,
  });
}
